src/selectors/profile.js

```diff
@@ -347,6 +347,151 @@ function _createCounterSelectors(counterIndex: CounterIndex) {
 export const getIPCMarkerCorrelations: Selector<IPCMarkerCorrelations> =
   createSelector(getThreads, correlateIPCMarkers);
 
+/**
+ * Returns an InnerWindowID -> Page map, so we can look up the page from inner
+ * window id quickly. Returns null if there are no pages in the profile.
+ */
+export const getInnerWindowIDToPageMap: Selector<Map<
+  InnerWindowID,
+  Page,
+> | null> = createSelector(getPageList, (pages) => {
+  if (!pages) {
+    // Return null if there are no pages.
+    return null;
+  }
+
+  const innerWindowIDToPageMap: Map<InnerWindowID, Page> = new Map();
+  for (const page of pages) {
+    innerWindowIDToPageMap.set(page.innerWindowID, page);
+  }
+
+  return innerWindowIDToPageMap;
+});
+
+/**
+ * Returns an InnerWindowID -> TabID map, so we can find the TabID of a given
+ * innerWindowID quickly. Returns null if there are no pages in the profile.
+ */
+export const getInnerWindowIDToTabMap: Selector<Map<
+  InnerWindowID,
+  TabID,
+> | null> = createSelector(
+  getPageList,
+  getInnerWindowIDToPageMap,
+  (pages, innerWindowIDToPageMap) => {
+    if (!pages || innerWindowIDToPageMap === null) {
+      // Return null if there are no pages.
+      return null;
+    }
+
+    const innerWindowIDToTabMap: Map<InnerWindowID, TabID> = new Map();
+    const getTopMostParent = (page) => {
+      if (page.embedderInnerWindowID === 0) {
+        return page;
+      }
+
+      // We are using a Map to make this more performant.
+      // It should be 1-2 loop iteration in 99% of the cases.
+      const parent = innerWindowIDToPageMap.get(page.embedderInnerWindowID);
+      if (parent !== undefined) {
+        return getTopMostParent(parent);
+      }
+      return page;
+    };
+    for (const page of pages) {
+      const topMostParent = getTopMostParent(page);
```

Comment:
It's not clear to me why we need to get the topmost parent, it looks like `page.tabID` should have the right value already, right?

~~And in that case, similarly to above, because we have a small array usually, we could do a simple loop over all pages instead of using a map and memoizing it.~~

I see it's used quite a lot in the selector below so why not after all.

Reply:
Ohh, that's a great point! I totally missed it that when I was implementing. Thanks!